server/channels/app/authentication.go

```diff
@@ -147,31 +147,70 @@ func (a *App) DoubleCheckPassword(rctx request.CTX, user *model.User, password s
 	return nil
 }
 
-func (a *App) checkLdapUserPasswordAndAllCriteria(rctx request.CTX, ldapId *string, password string, mfaToken string) (*model.User, *model.AppError) {
-	if a.Ldap() == nil || ldapId == nil {
+func (a *App) checkLdapUserPasswordAndAllCriteria(rctx request.CTX, userID, loginID, password, mfaToken string) (*model.User, *model.AppError) {
+	// MM-37585: Use locks to avoid concurrently checking AND updating the failed login attempts.
+	a.ch.ldapLoginAttemptsMut.Lock()
+	defer a.ch.ldapLoginAttemptsMut.Unlock()
+
+	user, appErr := a.GetUserForLogin(rctx, userID, loginID)
+	if appErr != nil {
+		return nil, appErr
+	}
+
+	ldapID := user.AuthData
+
+	if a.Ldap() == nil || ldapID == nil {
 		err := model.NewAppError("doLdapAuthentication", "api.user.login_ldap.not_available.app_error", nil, "", http.StatusNotImplemented)
 		return nil, err
 	}
 
-	ldapUser, err := a.Ldap().DoLogin(rctx, *ldapId, password)
+	// First time LDAP users will not have a userID
+	if user.Id != "" {
+		if err := checkUserLoginAttempts(user, *a.Config().LdapSettings.MaximumLoginAttempts); err != nil {
+			return nil, err
+		}
+	}
+
+	ldapUser, err := a.Ldap().DoLogin(rctx, *ldapID, password)
 	if err != nil {
 		// Log a info to make it easier to admin to spot that a user tried to log in with a legitimate user name.
 		if err.Id == "ent.ldap.do_login.invalid_password.app_error" {
-			rctx.Logger().LogM(mlog.MlvlLDAPInfo, "A user tried to sign in, which matched an LDAP account, but the password was incorrect.", mlog.String("ldap_id", *ldapId))
+			rctx.Logger().LogM(mlog.MlvlLDAPInfo, "A user tried to sign in, which matched an LDAP account, but the password was incorrect.", mlog.String("ldap_id", *ldapID))
+		}
+
+		var getUserByAuthErr *model.AppError
+		ldapUser, getUserByAuthErr = a.GetUserByAuth(ldapID, model.UserAuthServiceLdap)
+		if getUserByAuthErr != nil {
+			return nil, getUserByAuthErr
+		}
```

Comment:
Why are we fetching the user again here? If the user exists in MM, `GetUserForLogin` did find it. If it's a fresh user, this will fail, right?

Reply:
Added a check for this so we only do this for new users.